### PR TITLE
feat(esl-utils): add provider function support into `@prop` decorator

### PR DESCRIPTION
**Pull request by ala-n**

### CODE REVIEW

File: src/modules/esl-utils/decorators/prop.ts [hunk 1]
```diff
@@ -1,11 +1,32 @@
...
+
+/** Builds own property setter */
+function setter(name: string, readonly?: boolean) {
+  if (readonly) return (): void => void 0;
+  return function (value: any): void {
+    Object.defineProperty(this, name, {
+      value,
+      writable: true,
+      configurable: true
+    });
```
Comment (dshovchko) [inline src/modules/esl-utils/decorators/prop.ts:26]: Should it be enumerable?

File: src/modules/esl-utils/decorators/test/prop.test.ts [hunk 2]
```diff
@@ -1,22 +1,108 @@
...
+      class TestElement extends HTMLElement {
+        @prop({value: ''})
+        @attr()
+        public field: string;
+      }
+      new TestElement();
+    }).toThrowError(/own property/);
+  });
+
+  describe('Overriding @attr works fine', () => {
```
Comment (dshovchko) [inline src/modules/esl-utils/decorators/test/prop.test.ts:97]: This looks like an assertion, not a describe of the tests.
MB
```suggestion
  describe('Checking the @attr overriding', () => {
```

File: src/modules/esl-utils/decorators/test/prop.test.ts [hunk 3]
```diff
@@ -54,100 +140,82 @@ describe('Decorator: @prop', () => {
...
-    test('original decorator should not be executed', () => {
+    test('Original decorator should not be executed', () => {
       const el = new TestElement();
+      el.field = true;
       expect(el.getAttribute('field2')).toBe(null);
     });
   });
 
-  describe('Overriding @jsonAttr', () => {
+  describe('Overriding @jsonAttr with @prop works fine', () => {
```
Comment (dshovchko) [inline src/modules/esl-utils/decorators/test/prop.test.ts:170]: and here
```suggestion
  describe('Checking the @jsonAttr overriding', () => {
```

**Review by dshovchko — APPROVED** [review]

File: src/modules/esl-utils/decorators/test/prop.test.ts (hunk 2 above)
Comment (ala-n) [reply src/modules/esl-utils/decorators/test/prop.test.ts:97]: BTW: ideal test should have a Test Case name, so it is statement but not the status or action
So I'd say both of our versions are not descriptive enough. So let's keep that update for now.

File: src/modules/esl-utils/decorators/prop.ts (hunk 1 above)
Comment (ala-n) [reply src/modules/esl-utils/decorators/prop.ts:26]: That's a good (and tricky) question, by the way... 

We have an enumerable option for the original property, which is not a used characteristic that was broken in addition.
Note that even if we pass enumerable: true, we will not get it enumerable on the instance; we just make it such on the prototype.
So, I'm not sure we chose the right strategy to support it at all.

For instance, the override that we are watching for, I believe it's an even more realistic case to make it enumerable, but I don't think it's popular.

So let's create an issue to come back to that question...

File: src/modules/esl-utils/decorators/test/prop.test.ts (hunk 2 above)
Comment (dshovchko) [reply src/modules/esl-utils/decorators/test/prop.test.ts:97]: ok

**Review by abarmina — APPROVED** [review]

**Review by NastaLeo — APPROVED** [review]

**Pull request merged**